### ISSUE TITLE
SymbolTable.DoesMethodHaveParameterArray throws exception when accessing a dynamic created property

### ISSUE THREAD

**Issue by StefH** [dotnet/corefx#7596]
--> Tested with **dnxcore50**:

The following helper code to create a dynamic type with certain properties:

``` c#
public static dynamic Create(IDictionary<string, Type> properties)
{
    AssemblyName name = new AssemblyName("Test.DynamicClasses, Version=1.0.0.0");

    AssemblyBuilder assembly = AssemblyBuilder.DefineDynamicAssembly(name, AssemblyBuilderAccess.RunAndCollect);

    var module = assembly.DefineDynamicModule("Test.DynamicClasses");

    string typeName = "DynamicClass1";
    TypeBuilder tb = module.DefineType(typeName, TypeAttributes.Class | TypeAttributes.Public, typeof(DynamicClass));
    GenerateProperties(tb, properties);

    Type result = tb.CreateTypeInfo().AsType();

    return Activator.CreateInstance(result);
}
```

And the method which uses IL to create the backing fields, get/set methods and the properties:

``` c#
private static FieldInfo[] GenerateProperties(TypeBuilder tb, IDictionary<string, Type> properties)
{
    const MethodAttributes getSetAttr = MethodAttributes.Public | MethodAttributes.SpecialName | MethodAttributes.HideBySig;

    FieldInfo[] fields = new FieldBuilder[properties.Count];
    int i = 0;
    foreach (string key in properties.Keys)
    {
        string name = key;
        Type type = properties[key];
        FieldBuilder fb = tb.DefineField("_" + key, type, FieldAttributes.Public);
        PropertyBuilder pb = tb.DefineProperty(key, PropertyAttributes.HasDefault, type, null);

        MethodBuilder mbGet = tb.DefineMethod("get_" + name, getSetAttr, type, Type.EmptyTypes);
        ILGenerator genGet = mbGet.GetILGenerator(8);
        genGet.Emit(OpCodes.Ldarg_0);
        genGet.Emit(OpCodes.Ldfld, fb);
        genGet.Emit(OpCodes.Ret);
        pb.SetGetMethod(mbGet);

        MethodBuilder mbSet = tb.DefineMethod("set_" + name, getSetAttr, null, new Type[] { type });
        ILGenerator genSet = mbSet.GetILGenerator(8);
        genSet.Emit(OpCodes.Ldarg_0);
        genSet.Emit(OpCodes.Ldarg_1);
        genSet.Emit(OpCodes.Stfld, fb);
        genSet.Emit(OpCodes.Ret);
        pb.SetSetMethod(mbSet);

        fields[i] = fb;
    }

    return fields;
}       
```

And the test code is like:

``` c#
var list = new Dictionary<string, Type>();
list.Add("x", typeof(string));

var obj = Fact.Create(list);

obj._x = "ddd"; // accessing backing field directly works
string x_ = obj._x; // works
Console.WriteLine("x_=[{0}]", x_);

obj.x = "err"; // does not work --> throws exception like below

string x = obj.x; // does not work --> throws exception like below
Console.WriteLine(x);
```

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CSharp.RuntimeBinder.SymbolTable.DoesMethodHaveParameterArray(ParameterInfo[] parameters)
   at Microsoft.CSharp.RuntimeBinder.SymbolTable.AddMethodToSymbolTable(MemberInfo member, AggregateSymbol callingAggregate, MethodKindEnum kind)
   at Microsoft.CSharp.RuntimeBinder.SymbolTable.AddPropertyToSymbolTable(PropertyInfo property, AggregateSymbol aggregate)
   at Microsoft.CSharp.RuntimeBinder.SymbolTable.AddNamesInInheritanceHierarchy(String name, List`1 inheritance)
   at Microsoft.CSharp.RuntimeBinder.SymbolTable.PopulateSymbolTableWithName(String name, IEnumerable`1 typeArguments, Type callingType)
   at Microsoft.CSharp.RuntimeBinder.RuntimeBinder.PopulateSymbolTableWithPayloadInformation(DynamicMetaObjectBinder payload, Type callingType, ArgumentObject[] arguments)
   at Microsoft.CSharp.RuntimeBinder.RuntimeBinder.BindCore(DynamicMetaObjectBinder payload, IEnumerable`1 parameters, DynamicMetaObject[] args, DynamicMetaObject& deferredBinding)
   at Microsoft.CSharp.RuntimeBinder.RuntimeBinder.Bind(DynamicMetaObjectBinder payload, IEnumerable`1 parameters, DynamicMetaObject[] args, DynamicMetaObject& deferredBinding)
   at Microsoft.CSharp.RuntimeBinder.BinderHelper.Bind(DynamicMetaObjectBinder action, RuntimeBinder binder, IEnumerable`1 args, IEnumerable`1 arginfos, DynamicMetaObject onBindingError)
   at Microsoft.CSharp.RuntimeBinder.CSharpSetMemberBinder.FallbackSetMember(DynamicMetaObject target, DynamicMetaObject value, DynamicMetaObject errorSuggestion)
   at System.Dynamic.DynamicMetaObject.BindSetMember(SetMemberBinder binder, DynamicMetaObject value)
   at System.Dynamic.SetMemberBinder.Bind(DynamicMetaObject target, DynamicMetaObject[] args)
   at System.Dynamic.DynamicMetaObjectBinder.Bind(Object[] args, ReadOnlyCollection`1 parameters, LabelTarget returnLabel)
   at System.Runtime.CompilerServices.CallSiteBinder.BindCore[T](CallSite`1 site, Object[] args)
```

**Comment by StefH** [comment]
The issue is located in `Microsoft.CSharp.RuntimeBinder\SymbolTable.cs`:

```c#
private bool DoesMethodHaveParameterArray(ParameterInfo[] parameters)
{
	if (parameters.Length == 0)
	{
		return false;
	}

	ParameterInfo p = parameters[parameters.Length - 1];
	var attributes = p.GetCustomAttributes(false); // ISSUE is here !!!

	foreach (object o in attributes)
	{
		if (o is ParamArrayAttribute)
		{
			return true;
		}
	}
	return false;
}
```

The method `var attributes = p.GetCustomAttributes(false);` returns **null**.
However according to the specifications, it should return:
`A collection of the custom attributes that are applied to <paramref name="element"/>, or an empty collection if no such attributes exist.
`.

**Comment by StefH** [comment]
A work-around to get the original code in this issue working is:

```c#
MethodBuilder mbSet = tb.DefineMethod("set_" + name, getSetAttr, null, new Type[] { type });
mbSet.DefineParameter(1, ParameterAttributes.In, "_" + key);
```

**Comment by karelz** [comment]
@StefH sorry for the extra delay. Did you mean that Reflection is not working properly (`TypeInfo.GetCustomAttributes`)?

**Comment by karelz** [comment]
Next step: We need to confirm where the bug is and suggestion for a fix in MS.CSharp or in Reflection.

**Comment by JonHanna** [comment]
The issue would seem to lie in Reflection and/or Reflection.Emit, though perhaps at root in the relevant implementation of `Type`, but I can't reproduce it with:

```C#

[Fact]
public void DefaultParametersEmptyAttributes()
{
    AssemblyName name = new AssemblyName("Test.DynamicClasses, Version=1.0.0.0");

    AssemblyBuilder assembly = AssemblyBuilder.DefineDynamicAssembly(name, AssemblyBuilderAccess.RunAndCollect);

    var module = assembly.DefineDynamicModule("Test.DynamicClasses");

    string typeName = "DynamicClass1";
    TypeBuilder tb = module.DefineType(typeName, TypeAttributes.Class | TypeAttributes.Public, typeof(object));
    var field = tb.DefineField("_test", typeof(string), FieldAttributes.Private);
    var prop = tb.DefineProperty("Test", PropertyAttributes.HasDefault, typeof(string), null);
    var get = tb.DefineMethod("get_Test", MethodAttributes.Public | MethodAttributes.SpecialName | MethodAttributes.HideBySig, typeof(string), Type.EmptyTypes);
    var gen = get.GetILGenerator(8);
    gen.Emit(OpCodes.Ldarg_0);
    gen.Emit(OpCodes.Ldfld, field);
    gen.Emit(OpCodes.Ret);
    prop.SetGetMethod(get);
    var set = tb.DefineMethod("set_Test", MethodAttributes.Public | MethodAttributes.SpecialName | MethodAttributes.HideBySig, null, new []{typeof(string)});
    gen = set.GetILGenerator(8);
    gen.Emit(OpCodes.Ldarg_0);
    gen.Emit(OpCodes.Ldarg_1);
    gen.Emit(OpCodes.Stfld, field);
    gen.Emit(OpCodes.Ret);
    prop.SetSetMethod(set);
    var type = tb.CreateTypeInfo().AsType();
    Assert.Equal(0, type.GetProperty("Test").GetSetMethod().GetParameters()[0].GetCustomAttributes(false).Length);
    dynamic d = Activator.CreateInstance(type);
    d.Test = "abc";
    Assert.Equal("abc", d.Test);
}
```

Maybe I'm missing something in my attempt to repo, but perhaps also the issue was fixed as a side effect of something else.

**Comment by JonHanna** [comment]
@StefH can you confirm if this is still and issue in 2.0?

Thinking about it again, I suspect this may have been a bug with the relatively incomplete implementation of `Type` that .NET Core once had compared to now, and my inability to reproduce it is due to that being improved.

**Comment by StefH** [comment]
I also cannot confirm if it's solved in 2.0, I don't have the (test) code available.

I'll just close this issue and create a new one in case I see issues.

(At this moment, the original code from https://github.com/StefH/System.Linq.Dynamic.Core is not yet ported to 2.0)
